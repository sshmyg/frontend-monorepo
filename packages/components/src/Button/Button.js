import React, { useMemo } from 'react';

import styles from './Button.module.css';

export const Button = ({ className, type = 'button', ...restProps }) => {
  const actualizedClassName = useMemo(
    () => [styles.button, className].filter(Boolean).join(' '),
    [className],
  );

  return (
    <button
      data-testid="c-button"
      {...restProps}
      type={type}
      className={actualizedClassName}
    />
  );
};
