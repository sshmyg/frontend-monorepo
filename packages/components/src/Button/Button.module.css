.button {
  border: 1px solid peru;
  padding: 5px 10px;
  border-radius: 0;
  cursor: pointer;
  background-color: #fff;
  display: inline-flex;
}
