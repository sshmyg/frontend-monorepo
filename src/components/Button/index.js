import { createElement } from 'react';
import PropTypes from 'prop-types';

export default function Button({ elementType, children, ...restProps }) {
  if (elementType === 'button' && !restProps.type) {
    restProps.type = elementType;
  }

  return createElement(elementType, restProps, children);
}

Button.propTypes = {
  elementType: PropTypes.oneOfType([PropTypes.string, PropTypes.func]),
  type: PropTypes.string,
  children: PropTypes.node,
};

Button.defaultProps = {
  elementType: 'button',
};
