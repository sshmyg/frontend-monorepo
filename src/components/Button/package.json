{
  "name": "@components/Button",
  "version": "0.0.1",
  "license": "UNLICENSED"
}
