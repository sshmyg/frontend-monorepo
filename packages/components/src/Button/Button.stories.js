import React from 'react';

import { Button } from './Button';

export default { title: 'Button', component: Button };

const Template = (args) => <Button {...args} />;

export const WithText = Template.bind({});

WithText.args = {
  children: 'Button text',
};

export const WithEmoji = Template.bind({});

WithEmoji.args = {
  children: (
    <span role="img" aria-label="so cool">
      😀 😎 👍 💯
    </span>
  ),
};
