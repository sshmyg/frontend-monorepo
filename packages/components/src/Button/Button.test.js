import React from 'react';
import { render, fireEvent } from '@testing-library/react';

import { Button } from './Button';

describe('<Button />', () => {
  it('should render Button', () => {
    const { container } = render(<Button />);
    const button = container.querySelector('button');

    expect(button).toBeInTheDocument();
  });

  it('should render Button type button by default', () => {
    const { container } = render(<Button />);
    const button = container.querySelector('button');

    expect(button.type).toBe('button');
  });

  it('should render Button type submit', () => {
    const { container } = render(<Button type="submit" />);
    const button = container.querySelector('button');

    expect(button.type).toBe('submit');
  });

  it('should render Button type reset', () => {
    const { container } = render(<Button type="reset" />);
    const button = container.querySelector('button');

    expect(button.type).toBe('reset');
  });

  it('should render Button with text', () => {
    const testText = 'Hello world';
    const { getByText } = render(<Button>{testText}</Button>);
    const button = getByText(testText);

    expect(button).toBeInTheDocument();
  });

  it('should onClick works', () => {
    const handleClick = jest.fn();

    const { container } = render(<Button onClick={handleClick} />);
    const button = container.querySelector('button');

    fireEvent.click(button);

    expect(handleClick).toHaveBeenCalled();
  });

  it('should matches the snapshot', () => {
    const { asFragment } = render(<Button>Hello world</Button>);

    expect(asFragment).toMatchSnapshot();
  });
});
