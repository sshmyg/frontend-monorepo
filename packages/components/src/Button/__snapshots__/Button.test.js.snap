// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> should matches the snapshot 1`] = `[Function]`;
